x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/logical/And.java

```diff
@@ -36,4 +36,17 @@ public And swapLeftAndRight() {
     public Or negate() {
         return new Or(location(), new Not(location(), left()), new Not(location(), right()));
     }
+
+    @Override
+    public boolean nullable() {
+        if (left().nullable() && right().foldable() && right().fold() == Boolean.TRUE) {
+            return true;
+        }
+
+        if (right().nullable() && left().foldable() && left().fold() == Boolean.TRUE) {
```

Comment:
While it might be the logical equivalent, I find the current statement harder to understand.
The logic for AND is: if either side is FALSE, I'm _not_ nullable; otherwise I am if _either_ child is.
For OR it would: if either side is TRUE, I'm _not_ nullable, otherwise the same as above.
In both cases, if the constant checks, there's no nullability check on the children.

The method above checks for the nullability first and then for the folding for the opposite value of each side which I find counter intuitive.

Reply:
Let's take `OR` and the rule you propose:
```
if (Foldables.foldsAndMatches(right(), o -> o == Boolean.TRUE) ||
    Foldables.foldsAndMatches(left(), o -> o == Boolean.TRUE)) {
    return false;
}
return left().nullable() || right().nullable();
```
and the following option:
`<expr> OR null` => we return `true` as right is nullable.
Then the `FoldNull` will check that one of the children is nullable (the right one) and fold the whole expression to null event if the `<expr>` evaluates to `true`.